lib/fetch/index.js

```diff
@@ -122,130 +122,178 @@ class Fetch extends EE {
   }
 }
 
+const channels = {}
+
+try {
+  const diagnosticsChannel = require('diagnostics_channel')
+  channels.start = diagnosticsChannel.channel('undici:fetch:start')
+  channels.end = diagnosticsChannel.channel('undici:fetch:end')
+  channels.asyncStart = diagnosticsChannel.channel('undici:fetch:asyncStart')
+  channels.asyncEnd = diagnosticsChannel.channel('undici:fetch:asyncEnd')
+  channels.error = diagnosticsChannel.channel('undici:fetch:error')
+} catch {
+  channels.start = { hasSubscribers: false }
+  channels.end = { hasSubscribers: false }
+  channels.asyncStart = { hasSubscribers: false }
+  channels.asyncEnd = { hasSubscribers: false }
+  channels.error = { hasSubscribers: false }
+}
+
 // https://fetch.spec.whatwg.org/#fetch-method
-function fetch (input, init = undefined) {
+function fetch (input, init = {}) {
+  const context = { input, init }
+
   webidl.argumentLengthCheck(arguments, 1, { header: 'globalThis.fetch' })
 
   // 1. Let p be a new promise.
   const p = createDeferredPromise()
 
+  const callAsyncStartEnd = () => {
+    channels.asyncStart.publish(context)
+    channels.asyncEnd.publish(context)
+  }
+
   // 2. Let requestObject be the result of invoking the initial value of
   // Request as constructor with input and init as arguments. If this throws
   // an exception, reject p with it and return p.
   let requestObject
 
   try {
-    requestObject = new Request(input, init)
-  } catch (e) {
-    p.reject(e)
-    return p.promise
-  }
+    channels.start.publish(context)
+
+    try {
+      requestObject = new Request(input, init)
+    } catch (e) {
+      context.error = e
+      channels.error.publish(context)
+      p.reject(e)
+      return p.promise.then(() => {
+        callAsyncStartEnd()
+        delete context.error
```

Comment:
The `delete` here might hurt, shouldn't be better just set it as `null`?

Reply:
switched this to null